internal/characters/thoma/burst.go

```diff
@@ -78,56 +79,100 @@ func (c *char) Burst(p map[string]int) action.ActionInfo {
 	}
 }
 
-func (c *char) burstProc() {
-	// does not deactivate on death
+func (c *char) getN0Delay() int {
+	active := c.Core.Player.ActiveChar()
+	activeCharKey := active.Base.Key
+	// The character doesn't have an alt form
+	if data.PercentDelay5[activeCharKey] == -1 {
+		return data.PercentDelay5[activeCharKey]
+	}
+
+	if active.StatusIsActive(data.AltFormStatusKeys[activeCharKey]) {
+		return data.PercentDelay5AltForms[activeCharKey]
+	}
+
+	return data.PercentDelay5[activeCharKey]
+}
+
+func (c *char) burstStateDelayFuncGen(src int) func() {
+	return func() {
+		//ignore if on ICD
+		if c.StatusIsActive(burstICDKey) || c.Core.Player.CurrentState() != action.NormalAttackState || c.burstTickSrc != src {
+			return
+		}
+		//this should start a new ticker if not on ICD and state is correct
+		c.summonFieryCollapse()
+		c.Core.Log.NewEvent("thoma burst on state change", glog.LogCharacterEvent, c.Index).
+			Write("state", action.NormalAttackState).
+			Write("icd", c.StatusExpiry(burstICDKey))
+		c.burstTickSrc = c.Core.F
+		//use the hitlag affected queue for this
+		c.QueueCharTask(c.burstTickerFunc(c.Core.F), 60) //check every 1sec
+	}
+}
+
+func (c *char) burstStateHook() {
 	c.Core.Events.Subscribe(event.OnStateChange, func(args ...interface{}) bool {
```

Comment:
This hook can be changed to OnAttack. Then you can skip the animation state check in the subscription here. Same applies to Xingqiu.

OnAttack gets emitted here:
https://github.com/genshinsim/gcsim/blob/97e8e28c0e21004127ac33dddc62faa91331fff7/pkg/core/player/exec.go#L157
OnStateChange gets emitted here:
https://github.com/genshinsim/gcsim/blob/c184fb9382670f65c852e6a568bd51819fe0fe65/pkg/core/player/animation/animation.go#L87

As you can see, OnAttack gets emitted right before SetActionUsed gets called, which is where OnStateChange is emitted. You don't need info about the previous state/action, so OnAttack should also work and reduce the amount times this function gets called.